ReSwift/CoreTypes/Thunk.swift

```diff
@@ -0,0 +1,33 @@
+//
+//  Thunk.swift
+//  ReSwift
+//
+//  Created by Daniel Martín Prieto on 06/09/2018.
+//  Copyright © 2018 Benjamin Encz. All rights reserved.
+//
+
+import Foundation
+
+public struct Thunk<State: StateType>: Action {
+    let body: (_ dispatch: @escaping DispatchFunction, _ getState: @escaping () -> State?) -> Void
+    public init(body: @escaping (
+                _ dispatch: @escaping DispatchFunction,
+                _ getState: @escaping () -> State?) -> Void) {
+        self.body = body
+    }
+}
+
+public func createThunksMiddleware<State: StateType>() -> Middleware<State> {
```

Comment:
Instead of this verbose free function, what do you think about an extension to Store like `Store.thunkMiddleware() -> Middleware<State>`? Then the factory would be contained in the main type people interact with; and the State type would be known. Alternatively, `Thunk<State>.middleware()` is not a bad place, either, in my opinion. What do you think, @mjarvis @danielmartinprieto ?

Reply:
I tried the second option you suggest and finally went with the free function because I didn't want to mess with the `Thunk` type, as the middleware would latter consume thunks, and this could be misleading for people. I also prefer not to put the static function in the main type because, as we've seen, this is a completely opt-in utility. I like the free function because it's simple, and it will always be namespaced by the framework, anyway, but let me know if you still prefer one of the other options or even create a new type for it (overkill, IMO).

Reply:
It sure is an opt-in when the new module is used. I wouldn't mind extending the core types in useful ways. I would prefer the factory to be an extension to `Middleware`, but sadly that's not possible :) Free functions in Swift feel weird to me, but I don't have any good arguments 🤷‍♂️

Reply:
I also find free functions feels weird in Swift, I was wondering if it was like this because of the JS/Redux heritage or an actual benefit for ReSwift. I ended up creating a Reducer and Middleware protocol that wraps the free function in my example app.

Reply:
I didn't understand what you said about the reducer and the middleware protocol that wraps the free function. Could you elaborate?

Reply:
I use this in my code:
```
protocol Middleware {
    associatedtype State

    func handle(action: Action, with getState: (() -> State?)?, and dispatch: DispatchFunction?)
}

extension Middleware {
    func create() -> ReSwift.Middleware<State> {
        return { dispatch, getState in
            return { next in
                return { action in
                    next(action)
                    self.handle(action: action, with: getState, and: dispatch)
                }
            }
        }
    }
}
```
So it can be used like this:
```
struct AssistantMiddleware: Middleware {
    // MARK: Constants

    typealias State = AppState

    // MARK: Properties

    var assistant: AssistantService = WatsonAssistantService.shared

    // MARK: Actions handling

    func handle(action: Action, with getState: (() -> AppState?)?, and dispatch: DispatchFunction?) {
        switch action {
        case ChatView.Actions.connect:
            assistant.send(message: nil, with: dispatch)

        case let InputBarView.Actions.send(message):
            assistant.send(message: message, with: dispatch)

        default:
            break
        }
    }
}
```
And for the reducer 
```
protocol Reducer {
    associatedtype StateType

    func reduce(_ action: Action, _ state: StateType?) -> StateType
}
```
So it can be used like this:
```
struct AppReducer: Reducer {
    // MARK: Constants

    typealias StateType = AppState

    // MARK: Properties

    let chatReducer = ChatReducer()

    // MARK: Actions

    func reduce(_ action: Action, _ state: AppState?) -> AppState {
        return AppState(
            chatState: chatReducer.reduce(action, state?.chatState)
        )
    }
}
```
Then I can just create my store like this:
```
final class AppStore: Store<AppState> {
    // MARK: Singleton

    static let shared = AppStore()

    // MARK: Initialization
    
    private convenience init() {
        let appReducer = AppReducer()
        let middlewares = [
            AssistantMiddleware()
        ]
        
        self.init(reducer: appReducer.reduce, state: AppState.defaults, middleware: middlewares.map { $0.create() })
    }

    internal required init(reducer: @escaping ReSwift.Reducer<State>, state: State?, middleware: [ReSwift.Middleware<State>]) {
        super.init(reducer: reducer, state: state, middleware: middleware)
    }
}
```
I try to stay away as much as possible from free funcution and global variable. I prefer to encapsulate everthing in a struct or class and have a singleton in case I want to access some common instance easily. 
Using this my app delegate is empty and when I want to access the store I do AppStore.shared.